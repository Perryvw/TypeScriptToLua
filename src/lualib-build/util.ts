import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
