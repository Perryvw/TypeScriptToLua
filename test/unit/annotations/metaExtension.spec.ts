import {
    annotationDeprecated,
    extensionCannotConstruct,
    metaExtensionMissingExtends,
} from "../../../src/transformation/utils/diagnostics";
import * as util from "../../util";

test("MetaExtension", () => {
    const tsHeader = `
        declare class _LOADED {}
        declare namespace debug {
            function getregistry(): any;
        }
        /** @metaExtension */
        class LoadedExt extends _LOADED {
            public static test() {
                return 5;
            }
        }
    `;

    const luaResult = util.testModule`
       export default debug.getregistry()["_LOADED"].test();
    `
        .setTsHeader(tsHeader)
        .ignoreDiagnostics([annotationDeprecated.code])
        // TODO Cant use expectToMatchJsResult because above is not valid TS/JS
        .getLuaExecutionResult();

    expect(luaResult.default).toBe(5);
});

test("IncorrectUsage", () => {
    util.testModule`
        /** @metaExtension */
        class LoadedExt {
            public static test() {
                return 5;
            }
        }
    `.expectDiagnosticsToMatchSnapshot([metaExtensionMissingExtends.code, annotationDeprecated.code]);
});

test("DontAllowInstantiation", () => {
    util.testModule`
        declare class _LOADED {}
        /** @metaExtension */
        class Ext extends _LOADED {}
        const e = new Ext();
    `.expectDiagnosticsToMatchSnapshot([annotationDeprecated.code, extensionCannotConstruct.code]);
});
