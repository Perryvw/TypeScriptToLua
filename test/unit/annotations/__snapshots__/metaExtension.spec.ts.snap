// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DontAllowInstantiation: code 1`] = `
"require(\\"lualib_bundle\\");
local __meta___LOADED = debug.getregistry()._LOADED
e = __TS__New(Ext)"
`;

exports[`DontAllowInstantiation: diagnostics 1`] = `
"main.ts(4,9): warning TSTL: '@metaExtension' is deprecated and will be removed in a future update. Please update your code before upgrading to the next release, otherwise your project will no longer compile. See https://typescripttolua.github.io/docs/advanced/compiler-annotations#metaextension for more information.
main.ts(5,19): error TSTL: Cannot construct classes with '@extension' or '@metaExtension' annotation."
`;

exports[`IncorrectUsage: code 1`] = `
"function LoadedExt.test(self)
    return 5
end"
`;

exports[`IncorrectUsage: diagnostics 1`] = `
"main.ts(3,9): error TSTL: '@metaExtension' annotation requires the extension of the metatable class.
main.ts(3,9): warning TSTL: '@metaExtension' is deprecated and will be removed in a future update. Please update your code before upgrading to the next release, otherwise your project will no longer compile. See https://typescripttolua.github.io/docs/advanced/compiler-annotations#metaextension for more information."
`;
